File: apps/ios/Sources/Baozi/Models/SavedAppsStore.swift
import Foundation
import Observation

/// Generated UniFFI records don't carry Swift-only conformances. Lifting
/// `SavedApp` to `Identifiable` lets SwiftUI `.sheet(item:)`/`ForEach` use
/// its UUIDv4 as the stable identity.
extension SavedApp: Identifiable {}

/// Lightweight shared navigation signal that various surfaces emit and
/// the root navigation observes. `pendingOpenAppId` opens a saved-app
/// detail; `pendingConversationThreadId` pops back and pushes the origin
/// conversation (used by the "View Conversation" button on the detail view).
@MainActor
@Observable
final class SavedAppsNavigation {
    static let shared = SavedAppsNavigation()

    private(set) var pendingOpenAppId: String?
    private(set) var pendingConversationThreadId: String?

    private init() {}

    func requestOpen(appId: String) {
        pendingOpenAppId = appId
    }

    func consumeRequest() -> String? {
        let id = pendingOpenAppId
        pendingOpenAppId = nil
        return id
    }

    func requestConversation(threadId: String) {
        pendingConversationThreadId = threadId
    }

    func consumeConversationRequest() -> String? {
        let id = pendingConversationThreadId
        pendingConversationThreadId = nil
        return id
    }
}

/// Thin Swift wrapper around the Rust `saved_app_*` functions. Rust owns the
/// on-disk format (split `saved_apps.json` index + per-app html/state files)
/// and the `update_saved_app` orchestration; Swift just picks the directory
/// and exposes an observable, Swift-shaped API to the rest of the app.
///
/// Mutations that change metadata (`promote`, `rename`, `delete`,
/// `replaceHtml`) tickle the iCloud KVS bridge via
/// `notifyRustPreferencesChanged()`. State saves do not — the inner
/// `state_json` blob is intentionally kept out of the small KVS envelope.
@MainActor
@Observable
final class SavedAppsStore {
    static let shared = SavedAppsStore()

    private(set) var apps: [SavedApp] = []

    @ObservationIgnored private let directory: String = SavedAppsDirectory.path
    @ObservationIgnored private var saveStateDebouncers: [String: Task<Void, Never>] = [:]
    @ObservationIgnored private var pendingStateSaves: [String: PendingStateSave] = [:]

    private struct PendingStateSave {
        let stateJson: String
        let schemaVersion: UInt32
    }

    private init() {
        reload()
    }

    // MARK: - Metadata

    func reload() {
        apps = savedAppsList(directory: directory).apps
    }

    func app(id: String) -> SavedApp? {
        apps.first(where: { $0.id == id })
    }

    /// Slug → SavedApp resolution within the given origin thread. If `threadId`
    /// is nil, falls back to any app matching the slug (for legacy widgets
    /// surfaced without a known origin).
    func app(slug: String, threadId: String?) -> SavedApp? {
        if let threadId {
            if let match = apps.first(where: { $0.appId == slug && $0.originThreadId == threadId }) {
                return match
            }
        }
        return apps.first(where: { $0.appId == slug })
    }

    /// Apps whose `originThreadId` matches the given thread, sorted by
    /// most-recently-updated first. Empty when the thread has no saved apps.
    func appsForThread(_ threadId: String) -> [SavedApp] {
        apps
            .filter { $0.originThreadId == threadId }
            .sorted { $0.updatedAtMs > $1.updatedAtMs }
    }

    func getWithPayload(id: String) -> SavedAppWithPayload? {
        savedAppGet(directory: directory, appId: id)
    }

    @discardableResult
    func promote(
        title: String,
        widgetHTML: String,
        width: Double,
        height: Double,
        originThreadId: String?
    ) throws -> SavedApp {
        let app = try savedAppPromote(
            directory: directory,
            title: title,
            widgetHtml: widgetHTML,
            width: width,
            height: height,
            originThreadId: originThreadId
        )
        reload()
        CloudKVSBridge.shared.notifyRustPreferencesChanged()
        return app
    }

    @discardableResult
    func rename(id: String, title: String) throws -> SavedApp {
        let app = try savedAppRename(directory: directory, appId: id, title: title)
        reload()
        CloudKVSBridge.shared.notifyRustPreferencesChanged()
        return app
    }

    func delete(id: String) throws {
        try savedAppDelete(directory: directory, appId: id)
        // Drop any in-flight debounced save for this app.
        saveStateDebouncers.removeValue(forKey: id)?.cancel()
        pendingStateSaves.removeValue(forKey: id)
        reload()
        CloudKVSBridge.shared.notifyRustPreferencesChanged()
    }

    // MARK: - State (debounced)

    func loadState(id: String) -> SavedAppState? {
        savedAppLoadState(directory: directory, appId: id)
    }

    /// Debounced per-app_id state save. Coalesces bursts (e.g. from a
    /// dragging slider) into one write on a 250 ms trailing edge. Errors
    /// from the Rust layer (including `SavedAppError.StateTooLarge`) are
    /// logged and swallowed — the app keeps working; the oversized save
    /// is just rejected.
    func saveState(id: String, stateJson: String, schemaVersion: UInt32) {
        pendingStateSaves[id] = PendingStateSave(
            stateJson: stateJson,
            schemaVersion: schemaVersion
        )
        saveStateDebouncers[id]?.cancel()
        saveStateDebouncers[id] = Task { @MainActor [weak self] in
            try? await Task.sleep(nanoseconds: 250_000_000)
            guard let self, !Task.isCancelled else { return }
            self.flushStateSave(id: id)
        }
    }

    private func flushStateSave(id: String) {
        guard let pending = pendingStateSaves.removeValue(forKey: id) else { return }
        saveStateDebouncers.removeValue(forKey: id)
        do {
            _ = try savedAppSaveState(
                directory: directory,
                appId: id,
                stateJson: pending.stateJson,
                schemaVersion: pending.schemaVersion
            )
        } catch {
            LLog.warn(
                "saved_apps",
                "save_state failed",
                fields: ["app_id": id, "error": String(describing: error)]
            )
        }
    }

    // MARK: - Update flow

    enum UpdateError: Error, LocalizedError {
        case rustError(String)

        var errorDescription: String? {
            switch self {
            case .rustError(let message): return message
            }
        }
    }

    /// Ask the shared Rust client to regenerate this saved app's HTML using
    /// the active server. Rust internally inherits the origin thread's
    /// model / reasoning settings when that thread is still known to the
    /// store. When those settings are missing, Rust falls back to
    /// gpt-5.3-codex-spark, low reasoning, and fast service tier. It also
    /// forces full-access permissions for the edit. Returns the refreshed
    /// `SavedApp` metadata on success; throws
    /// `UpdateError.rustError` on failure.
    func requestUpdate(id: String, serverId: String, prompt: String) async throws -> SavedApp {
        let result = await AppModel.shared.client.updateSavedApp(
            serverId: serverId,
            directory: directory,
            appId: id,
            userPrompt: prompt
        )
        switch result {
        case .success(let app):
            reload()
            CloudKVSBridge.shared.notifyRustPreferencesChanged()
            return app
        case .error(let message):
            throw UpdateError.rustError(message)
        }
    }
}

File: apps/android/app/src/main/java/com/kris99/baozi/android/state/SavedAppsStore.kt
package com.kris99.baozi.android.state

import android.content.Context
import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import uniffi.codex_mobile_client.SavedApp
import uniffi.codex_mobile_client.SavedAppException
import uniffi.codex_mobile_client.SavedAppState
import uniffi.codex_mobile_client.SavedAppUpdateResult
import uniffi.codex_mobile_client.SavedAppWithPayload
import uniffi.codex_mobile_client.savedAppDelete
import uniffi.codex_mobile_client.savedAppGet
import uniffi.codex_mobile_client.savedAppLoadState
import uniffi.codex_mobile_client.savedAppPromote
import uniffi.codex_mobile_client.savedAppRename
import uniffi.codex_mobile_client.savedAppReplaceHtml
import uniffi.codex_mobile_client.savedAppSaveState
import uniffi.codex_mobile_client.savedAppsForThread
import uniffi.codex_mobile_client.savedAppsList

/**
 * Thin Kotlin wrapper around the Rust `saved_app_*` persistence surfaces and
 * [AppClient.updateSavedApp]. Rust owns the storage format, JSON cap, atomic
 * writes, and the update orchestration; this object only fronts them with a
 * reactive [StateFlow] and a per-id debouncer so a dragging slider inside a
 * WebView does not spam disk writes.
 */
object SavedAppsStore {
    private const val TAG = "SavedAppsStore"
    private const val SAVE_STATE_DEBOUNCE_MS = 250L

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val mutex = Mutex()
    private val debouncers = mutableMapOf<String, DebouncedSave>()

    private val _apps = MutableStateFlow<List<SavedApp>>(emptyList())
    val apps: StateFlow<List<SavedApp>> = _apps.asStateFlow()

    private data class DebouncedSave(
        var job: Job,
        var pendingStateJson: String,
        var pendingSchemaVersion: UInt,
    )

    suspend fun reload(context: Context) {
        val snapshot = withContext(Dispatchers.IO) {
            savedAppsList(SavedAppsDirectory.path(context))
        }
        _apps.value = snapshot.apps
    }

    suspend fun promote(
        context: Context,
        title: String,
        widgetHtml: String,
        width: Double,
        height: Double,
        originThreadId: String?,
    ): SavedApp {
        val app = withContext(Dispatchers.IO) {
            savedAppPromote(
                SavedAppsDirectory.path(context),
                title,
                widgetHtml,
                width,
                height,
                originThreadId,
            )
        }
        reload(context)
        return app
    }

    suspend fun rename(context: Context, appId: String, title: String): SavedApp {
        val app = withContext(Dispatchers.IO) {
            savedAppRename(SavedAppsDirectory.path(context), appId, title)
        }
        reload(context)
        return app
    }

    suspend fun delete(context: Context, appId: String) {
        withContext(Dispatchers.IO) {
            savedAppDelete(SavedAppsDirectory.path(context), appId)
        }
        cancelPendingSave(appId)
        reload(context)
    }

    suspend fun getWithPayload(context: Context, appId: String): SavedAppWithPayload? =
        withContext(Dispatchers.IO) {
            savedAppGet(SavedAppsDirectory.path(context), appId)
        }

    suspend fun loadState(context: Context, appId: String): SavedAppState? =
        withContext(Dispatchers.IO) {
            savedAppLoadState(SavedAppsDirectory.path(context), appId)
        }

    /**
     * Apps whose `origin_thread_id` matches [threadId], newest-updated first.
     * Used by the home-screen app-row takeover.
     */
    suspend fun appsForThread(context: Context, threadId: String): List<SavedApp> =
        withContext(Dispatchers.IO) {
            savedAppsForThread(SavedAppsDirectory.path(context), threadId)
        }

    /**
     * Resolve a saved app by its model-chosen slug ([SavedApp.appId]) within a
     * specific thread. Slugs are unique only within a thread — the same slug
     * in two threads is two independent apps.
     */
    suspend fun appForSlug(context: Context, slug: String, threadId: String): SavedApp? =
        appsForThread(context, threadId).firstOrNull { it.appId == slug }

    /**
     * Trailing-edge debounced save keyed by [appId]. A burst of rapid writes
     * from the WebView (slider drag, typing) coalesces into one disk write
     * [SAVE_STATE_DEBOUNCE_MS] after the last call.
     *
     * [SavedAppException.StateTooLarge] and other Rust-side validation errors
     * are logged and swallowed: the widget keeps working, the oversized save
     * is simply not persisted.
     */
    fun saveState(
        context: Context,
        appId: String,
        stateJson: String,
        schemaVersion: UInt,
    ) {
        val appContext = context.applicationContext
        scope.launch {
            mutex.withLock {
                val existing = debouncers[appId]
                if (existing != null) {
                    existing.pendingStateJson = stateJson
                    existing.pendingSchemaVersion = schemaVersion
                    existing.job.cancel()
                }
                val job = scope.launch {
                    delay(SAVE_STATE_DEBOUNCE_MS)
                    val (jsonToWrite, schema) = mutex.withLock {
                        val pending = debouncers.remove(appId) ?: return@withLock null
                        pending.pendingStateJson to pending.pendingSchemaVersion
                    } ?: return@launch
                    try {
                        savedAppSaveState(
                            MobilePreferencesDirectory.path(appContext),
                            appId,
                            jsonToWrite,
                            schema,
                        )
                    } catch (e: SavedAppException.StateTooLarge) {
                        Log.w(TAG, "saveState rejected (too large) appId=$appId", e)
                    } catch (e: SavedAppException) {
                        Log.w(TAG, "saveState failed appId=$appId", e)
                    } catch (e: Exception) {
                        Log.e(TAG, "saveState unexpected failure appId=$appId", e)
                    }
                }
                debouncers[appId] = DebouncedSave(job, stateJson, schemaVersion)
            }
        }
    }

    /**
     * Flush any pending debounced save for [appId] synchronously. Call this
     * from the detail view's onDispose so a quick tap-and-leave still
     * persists.
     */
    suspend fun flushPendingSave(context: Context, appId: String) {
        val pending = mutex.withLock {
            val existing = debouncers.remove(appId) ?: return@withLock null
            existing.job.cancel()
            existing.pendingStateJson to existing.pendingSchemaVersion
        } ?: return
        val (stateJson, schemaVersion) = pending
        try {
            withContext(Dispatchers.IO) {
                savedAppSaveState(
                    SavedAppsDirectory.path(context),
                    appId,
                    stateJson,
                    schemaVersion,
                )
            }
        } catch (e: Exception) {
            Log.w(TAG, "flushPendingSave failed appId=$appId", e)
        }
    }

    private suspend fun cancelPendingSave(appId: String) {
        mutex.withLock {
            debouncers.remove(appId)?.job?.cancel()
        }
    }

    suspend fun requestUpdate(
        context: Context,
        serverId: String,
        appId: String,
        prompt: String,
    ): SavedApp {
        val appModel = AppModel.shared
        val result = appModel.client.updateSavedApp(
            serverId,
            SavedAppsDirectory.path(context),
            appId,
            prompt,
        )
        return when (result) {
            is SavedAppUpdateResult.Success -> {
                reload(context)
                result.app
            }

            is SavedAppUpdateResult.Error ->
                throw IllegalStateException(result.message)
        }
    }

    /**
     * Direct HTML replacement escape hatch for tests or future bulk-import
     * flows. The normal update path is [requestUpdate].
     */
    suspend fun replaceHtml(
        context: Context,
        appId: String,
        widgetHtml: String,
        width: Double,
        height: Double,
    ): SavedApp {
        val app = withContext(Dispatchers.IO) {
            savedAppReplaceHtml(
                SavedAppsDirectory.path(context),
                appId,
                widgetHtml,
                width,
                height,
            )
        }
        reload(context)
        return app
    }
}
